src/main/java/build/buildfarm/worker/shard/Worker.java

```diff
@@ -790,7 +810,7 @@ long nextInterval(long now) {
 
               void registerIfExpired() {
                 long now = System.currentTimeMillis();
-                if (now >= workerRegistrationExpiresAt) {
+                if (now >= workerRegistrationExpiresAt && !isDeregistered) {
```

Comment:
oh interesting.  `workerRegistrationExpiresAt`.  I wonder if that means all workers have an eventual expiration time?

Reply:
Yes, that the i thing i plan to sync with you and Yuriy later.

Worker registration has a short lifespan, that's why it has to reregister itself in a loop as long as the worker is alive. Another benefit of it, besides the ones we can think of the other day, is that we don't need to deregister a worker when it gets killed because of errors or scale in. That means we do need a separate mechanism to inform a worker that it should stop reregistering itself.